nixos/modules/services/networking/shellhub-agent.nix

```diff
@@ -0,0 +1,89 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  cfg = config.services.shellhub-agent;
+in {
+
+  ###### interface
+
+  options = {
+
+    services.shellhub-agent = {
+
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to enable the ShellHub Agent daemon, which allows
+          secure remote logins.
+        '';
+      };
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.shellhub-agent;
+        defaultText = "pkgs.shellhub-agent";
+        description = ''
+          Which ShellHub Agent package to use.
+        '';
+      };
+
+      tenantId = mkOption {
+        type = types.str;
+        example = "ba0a880c-2ada-11eb-a35e-17266ef329d6";
+        description = ''
+          The tenant ID to use when connecting to the ShellHub
+          Gateway.
+        '';
+      };
+
+      server = mkOption {
+        type = types.str;
+        default = "https://cloud.shellhub.io";
+        description = ''
+          Server address of ShellHub Gateway to connect.
+        '';
+      };
+
+      privateKey = mkOption {
+        type = types.path;
+        default = "/var/lib/shellhub-agent/private.key";
+        description = ''
+          Location where to store the ShellHub Agent private
+          key.
+        '';
+      };
+    };
+  };
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+
+    systemd.services.shellhub-agent = {
+      description = "ShellHub Agent";
+
+      wantedBy = [ "multi-user.target" ];
+      requires = [ "local-fs.target" ];
+      wants = [ "network-online.target" ];
+      after = [
+        "local-fs.target"
+        "network.target"
+        "network-online.target"
+        "time-sync.target"
+      ];
+
+      environment.SERVER_ADDRESS = cfg.server;
+      environment.PRIVATE_KEY = cfg.privateKey;
+      environment.TENANT_ID = cfg.tenantId;
+
+      serviceConfig = {
+        Restart = "on-failure";
+        ExecStart = "${cfg.package}/bin/agent";
```

Comment:
This runs the service as root. Is this required? If it is, then I suggest explicitly setting `User = "root"` and adding a comment why it's necessary. It may also be possible to not require root, but just some special capabilities, in which case you could try `DynamicUser = true` and `AmbientCapabalities`/`CapabilityBoundingSet` (see other modules for examples).

Reply:
Thanks @Infinisil for the review. For now, we need to use root but I'll investigate if in the future we can drop the capabilities.

Reply:
You added a comment mentioning that it requires root, but it doesn't explain _why_ that's the case, which is what I'm looking for (probably just because it needs to log into arbitrary users?)

Reply:
It is an SSH server so yes, it needs to initialize sessions with different users. I assumed this was implicit by it being an SSH server. Should I make this explicit there?

Reply:
I added this explicitly there.